ksqldb-streams/src/main/java/io/confluent/ksql/execution/streams/TableSelectBuilder.java

```diff
@@ -56,20 +69,71 @@ private TableSelectBuilder() {
 
     final Named selectName = Named.as(StreamsUtil.buildOpName(queryContext));
 
+    final Optional<MaterializationInfo.Builder> matBuilder =  table.getMaterializationBuilder();
+
+    final boolean forceMaterialize = !matBuilder.isPresent();
```

Comment:
heuristic for determining if we should force a materializing at this node or not

Reply:
Not sure I follow. It looks like the heuristic is that if it's not materialized, materialize it.

It looks like we did ultimately go the route of simply materializing every SELECT node, with the assumption that there will always be exactly one of these, and it will be at the end of the topology. Is that right?

Reply:
The heuristic is that if we can't fetch a state store from upstream, then Materialize it

Reply:
Aaah. That makes more sense. Thanks!

Reply:
It is an indirect way of saying only the select nodes at the end of Joins would be materialized (therefore, "heuristic")

Reply:
Just so I understand, I assume that to be materialized, there must be two conditions met:
- Formats must be non-null, indicating that the plan came from a newer version of ksql which wants to materialize such a join (and not an old ksql which wouldn't have done this)
- The topology is such that there isn't a state store upstream already and so we would want one here.

Is that correct?

Reply:
yes, @AlanConfluent that's correct

Reply:
Sounds good to have the simple heuristic for now. We can discuss about if we want to be smarter with trade-offs between materialization and query serving in the near future.